src/database-postgresql.cpp

```diff
@@ -302,4 +314,328 @@ void Database_PostgreSQL::listAllLoadableBlocks(std::vector<v3s16> &dst)
+
+	prepareStatement("load_player_metadata",
+		"SELECT attr, value FROM player_metadata WHERE player = $1");
+
+}
+
+bool PlayerDatabasePostgreSQL::playerDataExists(const std::string &playername)
+{
+	verifyDatabase();
+
+	const char *values[] = { playername.c_str() };
+	PGresult *results = execPrepared("load_player", 1, values, false);
+
+	bool res = (PQntuples(results) > 0);
+	PQclear(results);
+	return res;
+}
+
+void PlayerDatabasePostgreSQL::savePlayer(RemotePlayer *player)
+{
+	PlayerSAO* sao = player->getPlayerSAO();
+	if (!sao)
+		return;
+
+	verifyDatabase();
+
+	v3f pos = sao->getBasePosition();
+	std::string pitch = ftos(sao->getPitch());
+	std::string yaw = ftos(sao->getYaw());
+	std::string posx = ftos(pos.X);
+	std::string posy = ftos(pos.Y);
+	std::string posz = ftos(pos.Z);
+	std::string hp = itos(sao->getHP());
+	std::string breath = itos(sao->getBreath());
+	const char *values[] = {
+		player->getName(),
+		pitch.c_str(),
+		yaw.c_str(),
+		posx.c_str(), posy.c_str(), posz.c_str(),
+		hp.c_str(),
+		breath.c_str()
+	};
+
+	const char* rmvalues[] = { player->getName() };
+	beginSave();
+
+	if (getPGVersion() < 90500) {
+		if (!playerDataExists(player->getName())) {
+			execPrepared("create_player", 8, values, true, false);
+		}
+		else {
+			execPrepared("update_player", 8, values, true, false);
+		}
+	}
+	else {
+		execPrepared("save_player", 8, values, true, false);
+	}
+
+	// Write player inventories
+	execPrepared("remove_player_inventories", 1, rmvalues);
+	execPrepared("remove_player_inventory_items", 1, rmvalues);
+
+	std::vector<const InventoryList*> inventory_lists = sao->getInventory()->getLists();
+	for (u16 i = 0; i < inventory_lists.size(); i++) {
+		const InventoryList* list = inventory_lists[i];
+		std::string name = list->getName(), width = itos(list->getWidth()),
+			inv_id = itos(i), lsize = itos(list->getSize());
+
+		const char* inv_values[] = {
+			player->getName(),
+			inv_id.c_str(),
+			width.c_str(),
+			name.c_str(),
+			lsize.c_str()
+		};
+		execPrepared("add_player_inventory", 5, inv_values);
+
+		for (u32 j = 0; j < list->getSize(); j++) {
+			std::ostringstream os;
+			list->getItem(j).serialize(os);
+			std::string itemStr = os.str(), slotId = itos(j);
+
+			const char* invitem_values[] = {
+				player->getName(),
+				inv_id.c_str(),
+				slotId.c_str(),
+				itemStr.c_str()
+			};
+			execPrepared("add_player_inventory_item", 4, invitem_values);
+		}
+	}
+
+	execPrepared("remove_player_metadata", 1, rmvalues);
+	const PlayerAttributes &attrs = sao->getExtendedAttributes();
+	for (PlayerAttributes::const_iterator it = attrs.begin(); it != attrs.end(); ++it) {
+		const char *meta_values[] = {
+			player->getName(),
+			it->first.c_str(),
+			it->second.c_str()
+		};
+		execPrepared("save_player_metadata", 3, meta_values);
+	}
+	endSave();
+}
+
+bool PlayerDatabasePostgreSQL::loadPlayer(RemotePlayer *player, PlayerSAO *sao)
+{
+	if (!sao)
+		return false;
```

Comment:
should this happens at any time in normal usage?
consider converting to `sanity_check`